File: Webserver/CGI/full_record.py
#!/usr/bin/env python3
"""
   It receives a record ID from a web client;
     it searches record in solr via solr's API  and
     it displays the full record.
"""
__author__= 'bernardo gomez'
__date__ = 'september 2020'



import requests
from requests.exceptions import HTTPError

import json
import sys
import os 
import cgi
from pathlib import Path
import re
import displaypage
import urllib.parse
import cgitb
import biblio
cgitb.enable()


def main():
  """
   It expects a configuration file on the command line.
   All error messages go to the standard error.

  """
  if len(sys.argv) < 2:
       print_error("System error (15)")
       print("No configuration file",file=sys.stderr)
       return

  cfg_file=sys.argv[1]

 # retrieve configuration variables.
  solr_host=""
  solr_path=""
  webserver_host=""
  cgi_path=""
  search_script=""
  full_record_script=""
  display_full_record=""
  form_file=""
  credentials=""

  failed=0
  if not Path(cfg_file).exists():
       print_error("System error(16)")
       failed+=1
       print("configuration file doesn't exist",file=sys.stderr)
       return 

  cfg_pattern=re.compile("(.*?)=(.*)")
  with open(cfg_file) as f:
       for line in f:
          line=line.rstrip("\n")
          m=cfg_pattern.match(line)
          if m:
              if m.group(1) == "solr_host":
                 solr_host=m.group(2)
              if m.group(1) == "solr_path":
                 solr_path=m.group(2)
              if m.group(1) == "webserver_host":
                 webserver_host=m.group(2)
              if m.group(1) == "cgi_path":
                 cgi_path=m.group(2)
              if m.group(1) == "search_script":
                 search_script=m.group(2)
              if m.group(1) == "full_record_script":
                 full_record_script=m.group(2)
              if m.group(1) == "display_full_record":
                 display_full_record=m.group(2)
              if m.group(1) == "form_file":
                 form_file=m.group(2)
              if m.group(1) == "credentials":
                 credentials=m.group(2)

  if not Path(display_full_record).exists():
       failed+=1
       print(display_full_record, " not found",file=sys.stderr)

  if not Path(form_file).exists():
       print(form_file, " not found",file=sys.stderr)
       failed+=1

  if search_script == '':
       failed+=1
       print("search_script undefined ",file=sys.stderr)

  if credentials == '':
       print("credentials undefined ",file=sys.stderr)
       failed+=1

  if failed > 0:
       print_error("System error(17)")
       print("One or more configuration variables are undefined. Suggestion: revise configuration file.",file=sys.stderr)
       return

######

##
  form = cgi.FieldStorage()
  if "id" not in form:
     print_form(form_file)
     return

  mms_id=form["id"].value

### record id must be an integer.
  test_input=str(mms_id)
  if not test_input.isnumeric(): 
          print_error("invalid request")
          print("Record ID is not an integer.",file=sys.stderr)
          return

  authorized_user,authorized_passwd=credentials.split("_|_")

  process_req(solr_host,solr_path,display_full_record,webserver_host,cgi_path,search_script,mms_id,authorized_user,authorized_passwd)
  return

def print_error(message):
     """
        It produces an HTML page with
        an error message when a system
        failure is detected.
     """
     print("Content-Type: text/html")    # HTML is following
     print()                             # blank line, end of headers
     print('<!DOCTYPE html>')
     print('<html lang="en-US">')
     print('<head>')
     print('<title>Error</title>')
     print('<meta charset="utf-8">')
     print('<meta name="viewport" content="width=device-width, initial-scale=1">')
     print('</head>')
     print('<body>')
     print("<H1>",message,"</H1>")
     print('</body>')
     print('</html>')
     return
   
def print_page(result_page):
    """
        It produces an HTML page with results
        based on the input string.
    """

    print("Content-Type: text/html")    # HTML is following
    print()                             # blank line, end of headers
    print(result_page)
    return

def print_form(form_file):
    """
        It produces an HTML page with webform
        based on the input file.
    """

    print("Content-Type: text/html")    # HTML is following
    print()                             # blank line, end of headers
    with open(form_file) as f:
       for line in f:
          print(line)
    return



def process_req(solr_host,solr_path,display_full_record,webserver_host,cgi_path,search_script,mms_id,authorized_user,authorized_passwd):

  """
    input parameters:
     - solr_host: solr hostname;
     - solr_path: a CGI alias;
     - display_full_record: file that serves as an HTML template;
     - webserver_host: hostname of the webserver that hosts these scripts;
     - cgi_path: cgi directory for these scripts;
     - search_script: name of cgi script that performs query;
     - mms_id: ALMA bibliographic record ID;
     - authorized_user: solr's user name;
     - authorized_passwd: user's password;

  """
  from requests.auth import HTTPBasicAuth

 ## solr API call
  try:
      search=solr_host+solr_path+'select?q=id:'+str(mms_id)
      try:
         response = requests.get(search,auth=HTTPBasicAuth(authorized_user,authorized_passwd),timeout=10)
         jsonResponse = response.json()
      except:
         print("connection to solr server failed ",file=sys.stderr)
         print_error("System error(19)")
         return


      if jsonResponse["responseHeader"]["status"]  != 0:
         error_msg=jsonResponse["error"]["msg"]
         print ("solr error ",error_msg,file=sys.stderr)
         print_error("System error(20)")
         return

# open full-record html file.
      full_rec_string=""
      with open(display_full_record) as f:
         for line in f:
            full_rec_string+=str(line)

      number_found=jsonResponse["response"]["numFound"]
      start_record=jsonResponse["response"]["start"]
       #print(len(jsonResponse["response"]["docs"]))
      record_count=len(jsonResponse["response"]["docs"])
      if record_count == 1:
### invoke displaypage class.
            full_record=displaypage.fullRecord(display_full_record)
###
##   displaying user query might create confusion here
            user_query=""
###
            page_string=full_record.full_page(jsonResponse,user_query,webserver_host,cgi_path,search_script)
            print_page(page_string)
            return

      if record_count == 0:
            print_error("Record not found.")
            print("record not found in solr:",str(mms_id),file=sys.stderr)
            return
  except HTTPError as http_err:
    print(f'HTTP error occurred: {http_err}',file=sys.stderr)
  except Exception as err:
    print(f'Other error occurred: {err}',file=sys.stderr)
  return

if __name__ == "__main__":
    main()

File: Webserver/CGI/go/fullRecord.go
package main
import "fmt"
import "log"
import "strings"
import "os"
import "bufio"
import "net/http"
import "regexp"
import "time"
import "io/ioutil"
import "net/http/cgi"
import "encoding/json"
import "tinylib/displaypage"

func replaceForm(inPage *regexp.Regexp,pageString string,userQuery string) string{
      line:=strings.Split(pageString,"\n")
      page:=""
      for _,val:= range line {
          result:=inPage.FindStringSubmatch(val)
          if result != nil {
	      page+=result[1]+result[2]+userQuery+result[4]+"\n"
           }  else {
              page+=val+"\n"
           }
      }
      return page
}

func  processReq(solrHost string,solrPath string,displayFullRecord string,webserverHost string,cgiPath string,fullRecordScript string,searchScript string,mmsID string,authorizedUser string,authorizedPasswd string) {
	l:=log.New(os.Stderr, "", 0)
        client := &http.Client{
        Timeout: time.Second * 10,
//              Timeout: time.Millisecond * 1,
        }
	reqUrl:=solrHost+solrPath+"select?q=id:"+mmsID
        method:="GET"
// l.Println("url:",reqUrl)
        req, err := http.NewRequest(method, reqUrl, nil)
        if err != nil {
                fmt.Printf("error type: %T\n", err)
                fmt.Printf("error value: %v\n", err)
                log.Fatal("Error reading response. ", err)
                printError("System error(26). NewRequest() failed.")
        }
        req.Header.Set("Cache-Control", "no-cache")
        req.Header.Set("Content-Type", "application/json")
//        user:="tinylib"
//        passwd:="euclid2020"
//        req.SetBasicAuth(user, passwd)
        req.SetBasicAuth(authorizedUser, authorizedPasswd)
        response, err := client.Do(req)
	if err != nil {
	      l.Println("connection to solr server failed.",err)
              printError("System error(23). request to solr failed.")
	      return 
	}
	if response.Status != "200 OK" {
	   if response.Status == "401" {
	      l.Println("connection to solr server failed. Bad credentials.")
              printError("System error(22). Bad credentials.")
	      return 
           } else {
	      l.Println("response status:",response.Status)
              printError("System error(24)")
   	      return
           }
	}
  	body, err := ioutil.ReadAll(response.Body)
	var jbody map[string] interface{}
	err1:=json.Unmarshal(body,&jbody)
	if err1 != nil {
             printError("System error(25). Not a json response.")
	     l.Println("couldn't parse json response")
	     return 
        }
	//l.Println("===>",jbody)
	header,ok:=jbody["responseHeader"].(map[string] interface{})
        if !ok   {
  	        printError("Error. couldn't parse json response.")
		return
	}
	if header["status"].(float64) != 0 {
  		l.Println("status:",header["status"],)
  	        printError("Error. solr invalid request")
		return
	}
	jresponse,ok:=jbody["response"].(map[string] interface{})
	if !ok {
  	        printError("FAILED jresponse!")
		return
	}
	numFound:=jresponse["numFound"].(float64)
	_=numFound
	jdocs,ok:=jresponse["docs"].([] interface{})
	if !ok {
   	     l.Println("solr didn't return docs")
  	     printError("FAILED docs!")
   	     return
	}
	recordCount:=len(jdocs)
	fullRecordStr:=""
	data,err:=ioutil.ReadFile(displayFullRecord)
	if err != nil {
	       l.Println("open failed",err)
  	       printError("System error(38)")
	       return
        }
        fullRecordStr=string(data)
	_=fullRecordStr
	_=recordCount
	if recordCount == 1 {
            fullRec:=displaypage.New(displayFullRecord)
	    pageString,err:=fullRec.FullPage(body,webserverHost,cgiPath,searchScript)
	    if err == nil {
		    printPage(pageString)
		    return
	    } else {
  	          printError("Full record display failed")
		  return
	    }
	}
        printError("Under construction.")
	return
}
// "facet_counts":{
//  "facet_queries":{},
//  "facet_fields":{
//    "pubDate":[
//      "1989",2,
//      "2004",2,
//      "1994",1,
//    "language":[
//      "english",5,
//      "chinese",1]},
//  "facet_ranges":{},
//  "facet_intervals":{},
//  "facet_heatmaps":{}}


func printPage(resultPage string){
    fmt.Println("Content-Type: text/html")    
    fmt.Println("") 
    fmt.Println(resultPage)
    return
}

func printForm(formFile string) {

    fmt.Println("Content-Type: text/html")   
    fmt.Println("")  
    data, err := ioutil.ReadFile(formFile)
    if err != nil {
       printError("System error(2)")
       return
    }
    fmt.Printf(string(data))
    return
}

func printError(message string) {
     fmt.Println("Content-Type: text/html")   
     fmt.Println("") 
     fmt.Println("<!DOCTYPE html>")
     fmt.Println("<html lang=\"en-US\">")
     fmt.Println("<head>")
     fmt.Println("<title>Error</title>")
     fmt.Println("<meta charset=\"utf-8\">")
     fmt.Println("<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">")
     fmt.Println("</head>")
     fmt.Println("<body>")
     fmt.Println("<H1>",message,"</H1>")
     fmt.Println("</body>")
     fmt.Println("</html>")
     return
}
func main() {
   l := log.New(os.Stderr, "", 0)
   argLength := len(os.Args[1:])
   if argLength == 0 {
      printError("Not ready yet")
      os.Exit(0)
   }
   _,err := os.Stat(os.Args[1])
   if os.IsNotExist(err) {
       printError("System error")
       l.Println(err)
       os.Exit(0)
   }
   file,err:=os.Open(os.Args[1])
   if err != nil {
       printError("System error(2)")
       os.Exit(0)
   }
   defer file.Close()
   //var param[] string
   scanner := bufio.NewScanner(file)
   scanner.Split(bufio.ScanLines)
   reConf,err:=regexp.Compile(`(.*?)=(.*)`)
   if err != nil {
       printError("System error(3)")
       os.Exit(0)
   }
   
     var solrHost=""
     var solrPath=""
     var webserverHost=""
     var cgiPath=""
     var searchScript=""
     var fullRecordScript=""
     var displayFullRecord=""
     var formFile=""
     var credentials=""
     failed:=0

   for scanner.Scan(){
      line:=scanner.Text()
      result:=reConf.FindStringSubmatch(line)
      if len(result)  < 1 {
         printError("System error(4)")
         os.Exit(0)
      }
            //  if m.group(1) == "display_results":
            //     display_results=m.group(2)
            //  if m.group(1) == "form_file":
            //     form_file=m.group(2)

      if  result[1] == "solr_host" {
          solrHost=result[2]
      }
      if result[1] == "solr_path" {
          solrPath=result[2]
      }
      if result[1] == "webserver_host" {
          webserverHost=result[2]
      }
      if result[1] == "cgi_path" {
          cgiPath=result[2]
      }
      if result[1] == "search_script" {
          searchScript=result[2]
      }
      if result[1] == "full_record_script" {
          fullRecordScript=result[2]
      }
      if result[1] == "display_full_record" {
          displayFullRecord=result[2]
      }
      if result[1] == "form_file" {
          formFile=result[2]
      }
      if result[1] == "credentials" {
          credentials=result[2]
      }

//      l.Println("regexp :",result[1],result[2])
   }
   if solrHost == "" {
         failed+=1
         l.Println("config param solr_host undefined")
         // printError("System error(6)")
         // os.Exit(0)
   }
   if solrPath == "" {
         failed+=1
         l.Println("config param solr_path undefined")
   }
   if webserverHost  == "" {
         failed+=1
         l.Println("config param webserver_host  undefined")
   }
   if cgiPath  == "" {
         failed+=1
         l.Println("config param cgi_path  undefined")
   }
   if searchScript  == "" {
         failed+=1
         l.Println("config param search_script undefined")
   }
   if fullRecordScript  == "" {
         failed+=1
         l.Println("config param full_record_script undefined")
   }
   if displayFullRecord  == "" {
         failed+=1
         l.Println("config param display_full_record undefined")
   }
   if formFile  == "" {
         failed+=1
         l.Println("form_file undefined")
   }
   if failed > 0 {
          printError("System error(7)")
          os.Exit(0)
   }
   if credentials  == "" {
         failed+=1
         l.Println("credentials undefined")
   }

   req,err:=cgi.Request()
   if err != nil {
          printError("Request failed")
          os.Exit(0)
   }
   err=req.ParseForm()
   if err != nil {
          printError("ParseForm failed")
          os.Exit(0)
   }
   if len(req.Form) == 0 {
	   printForm(formFile)
	   os.Exit(0)
   }
   mmsID:=req.Form.Get("id")
   if mmsID == "" {
	 printForm(formFile)
	 os.Exit(0)
   }
   subs,err:=regexp.Compile(`<|>|;|'|&`)
   mmsID=subs.ReplaceAllString(mmsID,"")
   authorization:=strings.Split(credentials,"_|_")
   authorizedUser:=authorization[0]
   authorizedPasswd:=authorization[1]
   processReq(solrHost,solrPath,displayFullRecord,webserverHost,cgiPath,fullRecordScript,searchScript,mmsID,authorizedUser,authorizedPasswd)

}
